spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HealthCheckAndDeleteStep.java

```diff
@@ -0,0 +1,153 @@
+import org.springframework.cloud.skipper.server.domain.AppDeployerData;
+import org.springframework.cloud.skipper.server.repository.DeployerRepository;
+import org.springframework.cloud.skipper.server.repository.ReleaseRepository;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Reponsible for checking the health of the latest deployed release, and then deleting
+ * those applications from the previous release. Executed asynchronously to avoid the
+ * client waiting a long time, potentially minutes, when executing update.
+ * @author Mark Pollack
+ */
+public class HealthCheckAndDeleteStep {
+
+	private final Logger logger = LoggerFactory.getLogger(SimpleRedBlackUpgradeStrategy.class);
+
+	private final ReleaseRepository releaseRepository;
+
+	private final DeployerRepository deployerRepository;
+
+	private final DeleteStep deleteStep;
+
+	private final HealthCheckProperties healthCheckProperties;
+
+	public HealthCheckAndDeleteStep(ReleaseRepository releaseRepository,
+			DeployerRepository deployerRepository,
+			DeleteStep deleteStep,
+			HealthCheckProperties healthCheckProperties) {
+		this.releaseRepository = releaseRepository;
+		this.deployerRepository = deployerRepository;
+		this.deleteStep = deleteStep;
+		this.healthCheckProperties = healthCheckProperties;
+	}
+
+	@Async("skipperThreadPoolTaskExecutor")
+	@Transactional
+	public CompletableFuture<Void> waitForNewAppsToDeploy(Release existingRelease,
+			AppDeployerData existingAppDeployerData,
+			List<String> applicationNamesToUpgrade, Release replacingRelease,
+			AppDeployerData replacingAppDeployerData) {
+		try {
+			logger.info("Waiting for apps in release {}-v{} to be healthy.", replacingRelease.getName(),
+					replacingRelease.getVersion());
+			boolean healthy = this.isHealthy(replacingRelease, replacingAppDeployerData);
+
+			if (healthy) {
+				logger.info("Apps in release {}-v{} are healthy.", replacingRelease.getName(),
+						replacingRelease.getVersion());
+				logger.info("Deleting changed applications from previous release {}-v{}",
+						existingRelease.getName(),
+						existingRelease.getVersion());
+
+				this.deleteStep.delete(existingRelease, existingAppDeployerData, applicationNamesToUpgrade);
+
+				// Update Status in DB
+				Status status = new Status();
+				status.setStatusCode(StatusCode.DEPLOYED);
+				replacingRelease.getInfo().setStatus(status);
+				replacingRelease.getInfo().setDescription("Upgrade complete");
+				this.releaseRepository.save(replacingRelease);
+				logger.info("Release {}-v{} has been DEPLOYED", replacingRelease.getName(),
+						replacingRelease.getVersion());
+			}
+			else {
+				logger.error("New release " + replacingRelease.getName() + " was not detected as healthy." +
+						"Keeping existing releases.");
+				Status status = new Status();
+				status.setStatusCode(StatusCode.DEPLOYED);
```

Comment:
Is the status supposed to be `DEPLOYED` even though the apps aren't healthy ?

Reply:
Yes, we got rid of 'deploying' status code.  deployed is the state where we asked the platform to deploy our apps.  Similar to the deleted comment about.  I'm not super thrilled with the skipper status and platform status, but I don't see a better way ATM

Reply:
Yes, we got rid of 'deploying' status code.  deployed is the state where we asked the platform to deploy our apps.  Similar to the deleted comment about.  I'm not super thrilled with the skipper status and platform status, but I don't see a better way ATM